Handle 429 rate limits a bit

lswainemoore opened the pull request:
We are rate limiting more in the API, which means we should handle it more gracefully here. Before, if we hit a rate limit, in the course of paginating through data, we would fail and bail. Now, it'll retry up to twice, waiting a minute in between.

This is Claude-y code, but it seems to work.

lswainemoore committed:
handle 429 rate limits a bit

dhhagan commented on quantaq/client.py:
```diff
@@ -138,7 +139,19 @@ def request(self, endpoint, verb=GET, params=None, **kwargs) -> Response:
+            if response.status_code != 429:
+                return response
+
+            if attempt < max_retries:
```
Maybe make the delay a kwarg that is passed through? Same with max_retries? But maybe we have a max max_retries? That way this is a bit more flexible if we were to change the limiting in the future?

dhhagan reviewed (changes requested):
This generally looks good. We'll need to do a few additional things before pushing including: 
  1. Fix the tests
  2. Update the docs
  3. Bump the version

We can deploy an alpha for a week or so and provide instructions on how to install an alpha from pip?

lswainemoore committed:
add an rate limit exceeded exception + basic test of rate-limiting in pagination context

lswainemoore committed:
try deleting this to get pytest to run in CI

lswainemoore committed:
try this now

lswainemoore committed:
just for now

lswainemoore committed:
update alpha version (manually)

lswainemoore committed:
429 settings via client kwargs, so that we can skip sleep during test + be more generally configurable

lswainemoore committed:
fix small mutable defaults bug

lswainemoore commented on quantaq/client.py:
```diff
@@ -90,6 +98,7 @@ def paginate(self, endpoint, verb, params, data):
                 params[k] = v
             
             # re-issue the request for the next page
+            # TODO should probably catch non-200 errors here.
```
note

lswainemoore commented on quantaq/client.py:
```diff
@@ -138,11 +147,26 @@ def request(self, endpoint, verb=GET, params=None, **kwargs) -> Response:
+
+        return response
+
+    def requests(self, endpoint, verb=GET, params=None, **kwargs):
         """Request, but for many of them (i.e. deals with pagination)
         """
+        params = dict() if params is None else params
```
this change fixes a small, unrelated mutable defaults bug

lswainemoore committed:
docs for rate limiting (seems to have picked up some solar stuff too)

lswainemoore commented:
@dhhagan I have done a few things since last review:

1) fix broken tests in CI
2) add a test of rate limiting behavior
3) made rate limiting params customizable via client base
4) updated docs (seemed to pick up changes for solar stuff when I ran `make html`)
5) fixed a small bug involving mutable defaults
6) made a small change actually throwing an error when 429 retries are exceeded
7) updated alpha (manually in pyrpoject.toml)
